src/passport-saml/xml-utils.ts

```diff
@@ -0,0 +1,119 @@
+  return values.every((value) => {
+    if (typeof value != "object") {
+      return false;
+    }
+    return typeof value.nodeType === "number" && value.nodeType === value.ATTRIBUTE_NODE;
+  });
+};
+
+const elementsXPathTypeGuard = (values: SelectedValue[]): values is Element[] => {
+  return values.every((value) => {
+    if (typeof value != "object") {
+      return false;
+    }
+    return typeof value.nodeType === "number" && value.nodeType === value.ELEMENT_NODE;
+  });
+};
+
+export const xpath = {
+  selectAttributes: (node: Node, xpath: string): Attr[] =>
+    selectXPath(attributesXPathTypeGuard, node, xpath),
+  selectElements: (node: Node, xpath: string): Element[] =>
+    selectXPath(elementsXPathTypeGuard, node, xpath),
+};
+
+export const decryptXml = async (xml: string, decryptionKey: string | Buffer) =>
+  util.promisify(xmlenc.decrypt).bind(xmlenc)(xml, { key: decryptionKey });
+
+const normalizeNewlines = (xml: string): string => {
+  // we can use this utility before passing XML to `xml-crypto`
+  // we are considered the XML processor and are responsible for newline normalization
+  // https://github.com/node-saml/passport-saml/issues/431#issuecomment-718132752
+  return xml.replace(/\r\n?/g, "\n");
+};
+
+// This function checks that the |signature| is signed with a given |cert|.
+export const validateXmlSignatureForCert = (
+  signature: Node,
+  certPem: string,
+  fullXml: string,
+  currentNode: Element
+): boolean => {
+  const sig = new xmlCrypto.SignedXml();
+  sig.keyInfoProvider = {
+    file: "",
+    getKeyInfo: () => "<X509Data></X509Data>",
+    getKey: () => Buffer.from(certPem),
+  };
+  const signatureStr = normalizeNewlines(signature.toString());
+  sig.loadSignature(signatureStr);
+  // We expect each signature to contain exactly one reference to the top level of the xml we
+  //   are validating, so if we see anything else, reject.
+  if (sig.references.length != 1) return false;
+  const refUri = sig.references[0].uri!;
+  const refId = refUri[0] === "#" ? refUri.substring(1) : refUri;
+  // If we can't find the reference at the top level, reject
+  const idAttribute = currentNode.getAttribute("ID") ? "ID" : "Id";
+  if (currentNode.getAttribute(idAttribute) != refId) return false;
+  // If we find any extra referenced nodes, reject.  (xml-crypto only verifies one digest, so
+  //   multiple candidate references is bad news)
+  const totalReferencedNodes = xpath.selectElements(
+    currentNode.ownerDocument,
+    "//*[@" + idAttribute + "='" + refId + "']"
+  );
+
+  if (totalReferencedNodes.length > 1) {
+    return false;
+  }
+  fullXml = normalizeNewlines(fullXml);
+  return sig.checkSignature(fullXml);
+};
+
+export const parseDomFromString = (xml: string): Document => {
+  return new xmldom.DOMParser().parseFromString(xml);
+};
+
+export const parseXml2JsFromString = async (xml: string | Buffer): Promise<any> => {
```

Comment:
Some improvements can be done on types here I think, but that's going to be for another time.